pkg/virt-controller/services/rendercontainer.go

```diff
@@ -0,0 +1,157 @@
+package services
+
+import (
+	k8sv1 "k8s.io/api/core/v1"
+
+	v1 "kubevirt.io/api/core/v1"
+)
+
+type ContainerSpecRenderer struct {
+	imgPullPolicy   k8sv1.PullPolicy
+	isPrivileged    bool
+	launcherImg     string
+	name            string
+	userID          int64
+	volumeDevices   []k8sv1.VolumeDevice
+	volumeMounts    []k8sv1.VolumeMount
+	resources       k8sv1.ResourceRequirements
+	liveninessProbe *k8sv1.Probe
+	readinessProbe  *k8sv1.Probe
+	ports           []k8sv1.ContainerPort
+	capabilities    *k8sv1.Capabilities
+	args            []string
+}
+
+type Option func(*ContainerSpecRenderer)
+
+func NewContainerSpecRenderer(containerName string, launcherImg string, imgPullPolicy k8sv1.PullPolicy, opts ...Option) *ContainerSpecRenderer {
+	computeContainerSpec := &ContainerSpecRenderer{
+		imgPullPolicy: imgPullPolicy,
+		launcherImg:   launcherImg,
+		name:          containerName,
+	}
+	for _, opt := range opts {
+		opt(computeContainerSpec)
+	}
+	return computeContainerSpec
+}
+
+func (ccr *ContainerSpecRenderer) Render(cmd []string) k8sv1.Container {
+	return k8sv1.Container{
+		Name:            ccr.name,
+		Image:           ccr.launcherImg,
+		ImagePullPolicy: ccr.imgPullPolicy,
+		SecurityContext: securityContext(ccr.userID, ccr.isPrivileged, ccr.capabilities),
+		Command:         cmd,
+		VolumeDevices:   ccr.volumeDevices,
+		VolumeMounts:    ccr.volumeMounts,
+		Resources:       ccr.resources,
+		Ports:           ccr.ports,
+		Env:             ccr.envVars(),
+		LivenessProbe:   ccr.liveninessProbe,
+		ReadinessProbe:  ccr.readinessProbe,
+		Args:            ccr.args,
+	}
+}
+
+func (ccr *ContainerSpecRenderer) envVars() []k8sv1.EnvVar {
+	if ccr.userID == 0 {
+		return nil
+	}
+	return xdgEnvironmentVariables()
+}
+
+func WithNonRoot(userID int64) Option {
+	return func(renderer *ContainerSpecRenderer) {
+		renderer.userID = userID
+	}
+}
+
+func WithPrivileged() Option {
+	return func(renderer *ContainerSpecRenderer) {
+		renderer.isPrivileged = true
+	}
+}
+
+func WithCapabilities(vmi *v1.VirtualMachineInstance) Option {
+	return func(renderer *ContainerSpecRenderer) {
+		renderer.capabilities = &k8sv1.Capabilities{
+			Add:  getRequiredCapabilities(vmi),
+			Drop: []k8sv1.Capability{CAP_NET_RAW},
+		}
+	}
+}
+
+func WithVolumeDevices(devices []k8sv1.VolumeDevice) Option {
+	return func(renderer *ContainerSpecRenderer) {
+		renderer.volumeDevices = devices
+	}
+}
+
+func WithVolumeMounts(mounts []k8sv1.VolumeMount) Option {
```

Comment:
A thought:

Maybe it's better to get an arbitrary amount of parameters rather than a slice?
Something like:
```go
func WithVolumeDevices(devices ...k8sv1.VolumeDevice) Option {...}
```

I think it can fit right for many cases, e.g. something like:
```go
myFirstVolume := Volume{....}
mySecondVolume := Volume{....}

// With the current function signature
computeContainerOpts := []Option{
    WithVolumeDevices([]Volume{
        myFirstVolume,
	mySecondVolume,
    }), ...}
```

Could turn into:
```go
myFirstVolume := Volume{....}
mySecondVolume := Volume{....}

// With arbitrary amount of parameters
computeContainerOpts := []Option{
    WithVolumeDevices(myFirstVolume, mySecondVolume),
    ...}
```

Also, a slice can be easily converted to arbitrary list (`WithVolumeDevices(volumeSlice...)`), but the opposite looks kinda ugly and too verbose.

The current signature also forces `func sidecarVolumeMount() []k8sv1.VolumeMount` to return a slice, although it returns only one volume mount.

WDYT?

Reply:
I agree, I prefer that signature. Thanks for bringing it to my attention.